### ISSUE TITLE
Run OCR on HEIC and other non standard image types

### ISSUE THREAD

**Issue by lfcnassif** [sepinf-inc/IPED#515]
Currently tesseract supports a few formats. Unsupported formats could be converted to png/bmp temporarily using imagemagick to allow OCRing them.

**Comment by tc-wleite** [comment]
@lfcnassif, can I take a look on this? Or you already started or have something in mind?

In one of the cases I am working with, there a lot of photos of bank receipts and documents, in HEIC format (extracted from an iPhone). 
It seems HEIC is becoming more common, so it would be nice to be able to run OCR in such images (and other non-standard formats like PSD and WEBP).

**Comment by lfcnassif** [comment]
Just a suggestion, tesseract supported types could be reviewed, maybe some supported type is not being sent to tesseract by OCRParser

**Comment by tc-wleite** [comment]
> Just a suggestion, tesseract supported types could be reviewed, maybe some supported type is not being sent to tesseract by OCRParser

Sure!

**Comment by tc-wleite** [comment]
Before starting to deal with tesseract itself, I collected samples from several cases, just to figure out which are the image formats we  have, and how IPED handles each format today.

```
MIMETYPE                   EXT    IPED SUPPORT
-------------------------- ------ -------------------------------------
image/aces                 EXR    ImageMagick	
image/bmp                  BMP    ImageIO+JAI	
image/emf                  EMF    ImageMagick
image/gif                  GIF    ImageIO
image/heic                 HEIC   ImageMagick
image/jp2                  JP2    ImageIO+JPEG2000(Opt) or ImageMagick
image/jpeg                 JPG    ImageIO
image/jpx                  JPX    ImageMagick
image/png                  PNG    ImageIO
image/tiff                 TIF    ImageIO+JAI
image/vnd.adobe.photoshop  PSD    ImageIO+TwelveMonkeys
image/vnd.wap.wbmp         WBMP   ImageIO+JAI
image/vnd.zbrush.dcx       DCX    ImageIO+JAI
image/vnd.zbrush.pcx       PCX    ImageIO+JAI
image/webp                 WEBP   ImageMagick
image/wmf                  WMF    ImageMagick (LibreOffice???)
image/x-cmu-raster         RAS    ImageMagick
image/x-jp2-codestream     J2C    ImageIO+JPEG2000(Opt) or ImageMagick
image/x-portable-bitmap    PBM    ImageIO+JAI (LibreOffice???)
image/x-portable-graymap   PGM    ImageIO+JAI
image/x-portable-pixmap    PPM    ImageIO+JAI
image/x-rgb                RGB    ImageMagick
image/x-xbitmap            XBM    ImageMagick
```

Not related to OCR, but I noticed that image/wmf and image/x-portable-bitmap are being handled by LibreOffice (in the Viewer) and this behavior was inherited by DocThumbsTask (it is generating thumbnails using LibreOffice).
I tested with all the images I have (a few PBM and thousands of WMF), and using the standard image handling (for thumbnails and in the viewer) worked fine and produced better results (the image viewer is more convenient/faster than LibreOffice viewer).

@lfcnassif, is there any particular reason to send these formats to LibreOffice?
I noticed in the code that other image formats (e.g. vnd.adobe.photoshop, emf) were at some point directed to LibreOffice viewer but now they are commented out (going to the image viewer).

**Comment by lfcnassif** [comment]
If I remember correctly, years ago libreoffice used to render at least WMF faster than imagemagick. But that was a long time ago, and with old versions of those tools. Render time could have changed with current versions of libreoffice and image magick. So I agree to change current defaults if results are better or faster, for sure!

**Comment by tc-wleite** [comment]
> If I remember correctly, years ago libreoffice used to render at least WMF faster than imagemagick. But that was a long time ago, and with old versions of those tools. Render time could have changed with current versions of libreoffice and image magick. So I agree to change current defaults if results are better or faster, for sure! Thanks!

Now that you mentioned I am starting to remember...
Anyway, this is a very minor detail, and not directly related to this issue.
I will double check before suggesting any changes (and create another issue if necessary).

About the OCR itself, my next step will be to test each format, to double check which are handled by tesseract. 
I already checked its documentation about supported input formats, but I think it is better to test in practice.
Another detail that I noticed is that some image mimetypes contain multiple "subformats" (related to used compression) (e.g. JPG, TIF, BMP). I tested each of them for thumbnail generation and viewer support, and will do the same for tesseract.

**Comment by lfcnassif** [comment]
Great, thank you @tc-wleite for all your work. We also plan to upgrade tesseract on #206, not sure if supported formats have changed...

**Comment by tc-wleite** [comment]
> Great, thank you @tc-wleite for all your work. We also plan to upgrade tesseract on #206, not sure if supported formats have changed...

I saw #206. I plan to test with both versions, to see if there are any differences regarding supported input formats.

**Comment by tc-wleite** [comment]
Created a new issue (#575) with minor enhancements mostly related to non-standard image formats.
Some of the proposed changes will be useful for the conversion required to run the OCR in formats not natively supported by Tesseract.

**Comment by tc-wleite** [comment]
Back to OCR itself, I checked which formats are supported by Tesseract 4.0.0 (will check later for 5.x) and currently send by IPED to OCR (in OCRParser source code).

```
MIMETYPE                   EXT    IPED  TESSERACT 4 
-------------------------- -----  ----  -----------
image/aces                 EXR     -         -
image/bmp                  BMP    Yes	  Partial*
image/emf                  EMF     -         -
image/gif                  GIF    Yes       Yes    
image/heic                 HEIC    -         -
image/jp2                  JP2    Yes        -
image/jpeg                 JPG    Yes       Yes
image/jpx                  JPX    Yes        -
image/png                  PNG    Yes       Yes
image/svg+xml              SVG     -         -
image/tiff                 TIF    Yes       Yes
image/vnd.adobe.photoshop  PSD     -         -
image/vnd.wap.wbmp         WBMP    -         -
image/vnd.zbrush.dcx       DCX     -         -
image/vnd.zbrush.pcx       PCX     -         -
image/webp                 WEBP    -         -
image/wmf                  WMF     -         -
image/x-cmu-raster         RAS     -         -
image/x-jp2-codestream     J2C     -         -
image/x-portable-bitmap    PBM    Yes       Yes
image/x-portable-graymap   PGM     -        Yes
image/x-portable-pixmap    PPM     -        Yes
image/x-rgb                RGB     -         -
image/x-xbitmap            XBM     -         -

* BMPs with compression (bi_rle4, bi_rle8, bi_bitfields) are not supported. 
```

**Comment by lfcnassif** [comment]
Thanks Wladimir, about JPX and JP2, I just copied them from Tika OCR code (originally based on our OCR code but being improved since them...)

**Comment by tc-wleite** [comment]
> Thanks Wladimir, about JPX and JP2, I just copied them from Tika OCR code (originally based on our OCR code but being improved since them...)

I see. Well, these file formats are not common, and I tested with a very small set (~10 of each). Maybe it would open other files. 

Anyway, whenever there is a "partial" support, probably is better to convert to a standard format, before sending to Tesseract.

Another detail, for some files there was a message that leptonica needs to be compiled with some specific option. 
I will investigate further, but it is possible that the version used in Windows (included in the tools f

**Comment by tc-wleite** [comment]
Confirmed here that currently used tesseract wasn't compiled with all possible libraries that handle image formats:

```
tesseract 4.0.0
 leptonica-1.76.0 (Jan 23 2019, 10:44:57) [MSC v.1900 LIB Release x64]
  libgif 5.1.4 : libjpeg 6b (libjpeg-turbo 1.5.3) : libpng 1.6.36 : libtiff 4.0.10 : zlib 1.2.11
 Found AVX
 Found SSE 
```

There a few other formats that could be directly supported by Tesseract.
I will try to compile it with all supported formats.
After that, not sure which is the best approach:

1. Assume that used Tesseract supports certain formats (for Windows this could be controlled because it is include in the "tools" already), and send just this formats directly. Basically, this is the current approach.
2. Check Tesseract's output for certain error messages (that seem to vary a lot), and in these cases convert the image to a standard format and send it again.
3. When checking Tesseract version, also check which libraries it includes (e.g. libgif, libjpeg), and send directly only images of such formats.

@lfcnassif, what do you think?
The first option would be the simplest, but the third one seems easy enough to implement, and would add some robustness to the process.

**Comment by lfcnassif** [comment]
Thanks @tc-wleite, probably my fault when compiling without OpenMP support, see: https://github.com/tesseract-ocr/tesseract/issues/1662#issuecomment-456991138

I think option 3 is good. If you are going to replace current tesseract, I think we could upgrade to version 5, seems faster and more accurate (#206), and maybe compiling it from source is not needed: https://github.com/UB-Mannheim/tesseract/wiki

**Comment by lfcnassif** [comment]
I didn't upgrade yet just because I would like to try to reduce the binary package size (much larger than tesseract 4) and I had some trouble to make it work in a portable way, but I didn't investigate it properly to try to fix the portability

**Comment by tc-wleite** [comment]
> I didn't upgrade yet just because I would like to try to reduce the binary package size (much larger than tesseract 4) and I had some trouble to make it work in a portable way, but I didn't investigate it properly to try to fix the portability

Ok, I will try to take a look on this too.

**Comment by lfcnassif** [comment]
I just checked the UB-Mannheim build --version output:
```
tesseract v5.0.0-alpha.20210506
 leptonica-1.78.0
  libgif 5.1.4 : libjpeg 8d (libjpeg-turbo 1.5.3) : libpng 1.6.34 : libtiff 4.0.9 : zlib 1.2.11 : libwebp 0.6.1 : libopenjp2 2.3.0
 Found AVX512BW
 Found AVX512F
 Found AVX2
 Found AVX
 Found FMA
 Found SSE4.1
 Found libarchive 3.5.0 zlib/1.2.11 liblzma/5.2.3 bz2lib/1.0.6 liblz4/1.7.5 libzstd/1.4.5
 Found libcurl/7.77.0-DEV Schannel zlib/1.2.11 zstd/1.4.5 libidn2/2.0.4 nghttp2/1.31.0
```

**Comment by tc-wleite** [comment]
I managed to build tesseract 5.0.0 from the source (for Windows), excluding things we don't need (or want) to use (like OpenMP and training related stuff), with the latest version of leptonica libraries, using a static build and minimal binary size (only ~4 MB). 

I will run a larger set of tests to compare performance, stability and the text recognition results, with the current version we use (4.0.0),  the UB-Mannheim build and this one I just built.

```
tesseract 5.0.0-alpha-20210401-130-g7a308
 leptonica-1.80.0 (Jun  3 2021, 17:33:07) [MSC v.1929 LIB Release x64]
  libgif 5.2.1 : libjpeg 9d : libpng 1.6.37 : libtiff 4.1.0 : zlib 1.2.11 : libwebp 1.1.0 : libopenjp2 2.4.0
 Found AVX512BW
 Found AVX512F
 Found AVX2
 Found AVX
 Found FMA
 Found SSE4.1
```

**Comment by lfcnassif** [comment]
> I managed to build tesseract 5.0.0 from the source (for Windows), excluding things we don't need (or want) to use (like OpenMP and training related stuff), with the latest version of leptonica libraries, using a static build and minimal binary size (only ~4 MB).

Great @tc-wleite! This is a really small static binary size!

**Comment by tc-wleite** [comment]
Running with the newer build,  **image/jp2** and **image/jpx** images were processed by tesseract.
Anyway, I will implement the third option described above, to check what libraries were included in the tesseract being used.
Only unsupported formats will be converted before sending to tesseract.
BMP will need a special handling, as tesseract only support uncompressed BMPs (so mimetype won't be enough to decide if external conversion is necessary).

**Comment by tc-wleite** [comment]
@lfcnassif, to implement this feature I will need to use ImageUtil and GraphicsMagicConverter classes inside OCRParser (**iped-parsers**), right?

The problem is that both these utilities classes are in the **iped-viewer** subproject.
They are already used in **iped-engine** subproject (by tasks like ImageThumbTask).

Wouldn't it make more sense to move these classes to somewhere else? Maybe **iped-utils**?
Is there any better/easier option?

**Comment by tc-wleite** [comment]
```
MIMETYPE                   EXT    IPED  TESSERACT 4  NEW IPED+TESSERACT
-------------------------- -----  ----  ----------- -------------------
image/aces                 EXR     -         -       Conv. ImageMagick
image/bmp                  BMP    Yes	  Partial*     Direct/ImageIO
image/emf                  EMF     -         -       Conv. ImageMagick
image/gif                  GIF    Yes       Yes            Direct
image/heic                 HEIC    -         -       Conv. ImageMagick
image/jp2                  JP2    Yes        -             Direct
image/jpeg                 JPG    Yes       Yes            Direct
image/jpx                  JPX    Yes        -             Direct
image/png                  PNG    Yes       Yes            Direct
image/svg+xml              SVG     -         -       Conv. ImageMagick
image/tiff                 TIF    Yes       Yes           Direct
image/vnd.adobe.photoshop  PSD     -         -         Conv. ImageIO
image/vnd.wap.wbmp         WBMP    -         -         Conv. ImageIO
image/vnd.zbrush.dcx       DCX     -         -         Conv. ImageIO
image/vnd.zbrush.pcx       PCX     -         -         Conv. ImageIO    
image/webp                 WEBP    -         -            Direct
image/wmf                  WMF     -         -       Conv. ImageMagick
image/x-cmu-raster         RAS     -         -       Conv. ImageMagick
image/x-jp2-codestream     J2C     -         -       Conv. ImageMagick
image/x-portable-bitmap    PBM    Yes       Yes           Direct
image/x-portable-graymap   PGM     -        Yes           Direct
image/x-portable-pixmap    PPM     -        Yes           Direct
image/x-rgb                RGB     -         -       Conv. ImageMagick
image/x-xbitmap            XBM     -         -       Conv. ImageMagick
```

**Comment by lfcnassif** [comment]
> @lfcnassif, to implement this feature I will need to use ImageUtil and GraphicsMagicConverter classes inside OCRParser (**iped-parsers**), right?
> 
> The problem is that both these utilities classes are in the **iped-viewer** subproject.
> They are already used in **iped-engine** subproject (by tasks like ImageThumbTask).
> 
> Wouldn't it make more sense to move these classes to somewhere else? Maybe **iped-utils**?
> Is there any better/easier option?

I would like to keep iped-util with as few dependencies as possible. Does ImageUtil need metadataextractor? If no, ok. If yes, maybe moving to iped-parsers is an option. GMConverter could go to iped-util I think it doesn't need external libs, right? And as it is being moved, it could be renamed to ImageMagickConverter, we switched to IM years ago.

**Comment by lfcnassif** [comment]
> ```
> MIMETYPE                   EXT    IPED  TESSERACT 4  NEW IPED+TESSERACT
> -------------------------- -----  ----  ----------- -------------------
> image/aces                 EXR     -         -       Conv. ImageMagick
> image/bmp                  BMP    Yes	  Partial*     Direct/ImageIO
> image/emf                  EMF     -         -       Conv. ImageMagick
> image/gif                  GIF    Yes       Yes            Direct
> image/heic                 HEIC    -         -       Conv. ImageMagick
> image/jp2                  JP2    Yes        -             Direct
> image/jpeg                 JPG    Yes       Yes            Direct
> image/jpx                  JPX    Yes        -             Direct
> image/png                  PNG    Yes       Yes            Direct
> image/svg+xml              SVG     -         -       Conv. ImageMagick
> image/tiff                 TIF    Yes       Yes           Direct
> image/vnd.adobe.photoshop  PSD     -         -         Conv. ImageIO
> image/vnd.wap.wbmp         WBMP    -         -         Conv. ImageIO
> image/vnd.zbrush.dcx       DCX     -         -         Conv. ImageIO
> image/vnd.zbrush.pcx       PCX     -         -         Conv. ImageIO    
> image/webp                 WEBP    -         -            Direct
> image/wmf                  WMF     -         -       Conv. ImageMagick
> image/x-cmu-raster         RAS     -         -       Conv. ImageMagick
> image/x-jp2-codestream     J2C     -         -       Conv. ImageMagick
> image/x-portable-bitmap    PBM    Yes       Yes           Direct
> image/x-portable-graymap   PGM     -        Yes           Direct
> image/x-portable-pixmap    PPM     -        Yes           Direct
> image/x-rgb                RGB     -         -       Conv. ImageMagick
> image/x-xbitmap            XBM     -         -       Conv. ImageMagick
> ```

Great! Does default ImageIO handle PCX and DCX or you are adding a new plugin?

**Comment by tc-wleite** [comment]
> > @lfcnassif, to implement this feature I will need to use ImageUtil and GraphicsMagicConverter classes inside OCRParser (**iped-parsers**), right?
> > The problem is that both these utilities classes are in the **iped-viewer** subproject.
> > They are already used in **iped-engine** subproject (by tasks like ImageThumbTask).
> > Wouldn't it make more sense to move these classes to somewhere else? Maybe **iped-utils**?
> > Is there any better/easier option?
> 
> I would like to keep iped-util with as few dependencies as possible. Does ImageUtil need metadataextractor? If no, ok. If yes, maybe moving to iped-parsers is an option. GMConverter could go to iped-util I think it doesn't need external libs, right? And as it is being moved, it could be renamed to ImageMagickConverter, we switched to IM years ago.

Ok, I got the idea now! 

About ImageUtil, some methods does depend on MetadataExtractor, but I think they could be separated from the rest, keeping only parts with no dependencies in iped-utils.

About GMConverter, I need to check in more detail, but its initialization may be tricky. 
It uses static methods to set important parameters, called by ImageThumbTask.
For using inside a parser, wouldn't be better to use something like the OCRParser does (through properties)?

I considered renaming this class. 
What about a more generic name (like ExternalImageConverter), as it still supports GM?

**Comment by tc-wleite** [comment]
> Great! Does default ImageIO handle PCX and DCX or you are adding a new plugin?

I will need to double check this next week (my working machine seems to be offline), but I **think** it is supported because of JAI (jai-imageio-core), which may be present in optional_jars folder, right?

Although I listed as "ImageIO" conversion, in practice it will use the "standard procedure", first try the "internal" (ImageIO) conversion, then try "external" if necessary. 
There won't be anything hard coded regarding when to use internal/external conversion. 

There will be rules (some basic formats always included, plus other formats depending on Tesseract libraries available) to decide when to send it directly to Tesseract (as it is today) and when to convert before sending (only for formats not supported by Tesseract). The idea is to keep the current behavior for most common formats, as use the extra conversion step only for non-standard formats.

**Comment by lfcnassif** [comment]
> For using inside a parser, wouldn't be better to use something like the OCRParser does (through properties)?

Yes! Properties are used because of external parsing (enabledExternalParsing = true), the main process copies its properties to child parsing processes.

> What about a more generic name (like ExternalImageConverter), as it still supports GM?

Good!

**Comment by lfcnassif** [comment]
> I couldn't find an explicit mention to this dependency, so I guess it comes with some other library used in the project. (PDFBox?)

It comes from Tika

**Comment by lfcnassif** [comment]
Closed by #593. Thanks @tc-wleite great work.

**Comment by lfcnassif** [comment]
Hi @tc-wleite. @FelipeFcosta implemented some tests for OCRParser. Now he is trying to install tesseract and other external tools in the remote build environment, but then the build fails with:
`No X11 DISPLAY variable was set, but this program performed an operation which requires it.`

We think the possible cause is the `getCompatibleImage()` method calls in OCRParser. Do you think it is safe to remove those calls?

**Comment by tc-wleite** [comment]
> We think the possible cause is the `getCompatibleImage()` method calls in OCRParser. Do you think it is safe to remove those calls?

Good question... I wondered about that myself in the past.
As far as I know, creating these compatible images makes sense when they are going to be displayed, which is not the case of the OCRParser.
So I think it is possible (desirable?) to remove it. 
I would test it though, as I am not sure about possible side effects.

**Comment by lfcnassif** [comment]
@FelipeFcosta, please, comment out those calls and check if tests pass. One detail I forgot to say, if you need to change any code outside test classes, like the OCRParser if we apply this change, we should open an issue and implement the change in a separate PR different than the future tests PR.

**Comment by FelipeFcosta** [comment]
Tests still pass locally after commenting out `getCompatibleImage()` calls, and the error has disappeared in the remote build, but some of the test assertions fail in the remote. I guess it's because the tesseract version (5.0.0-alpha) installed on windows is more recent and can pick up more characters from the image.

**Comment by FelipeFcosta** [comment]
Should I make the assertions for characters that tesseract 4.0.0 can pick up on?

**Comment by lfcnassif** [comment]
> Should I make the assertions only for characters that tesseract 4.0.0 can pick up on?

Have you installed the portuguese language pack in remote CI? If yes, is it possible to install tesseract 5.0 in remote?

**Comment by lfcnassif** [comment]
> We think the possible cause is the `getCompatibleImage()` method calls in OCRParser

I just confirmed with @arisjr that OCR on TIF images with > 3 pages is not working in their processing environment (without X). I'll open an issue to fix it.

**Comment by lfcnassif** [comment]
> Should I make the assertions only for characters that tesseract 4.0.0 can pick up on?

@FelipeFcosta as I said to you, just to document here, I think we should change the OCR test images for better quality ones. Installing tesseract-5.0 on remote CI should fix the remote builds, but if an user tries to build the project in an environment with a different tesseract version, the build may fail in his machine...

**Comment by lfcnassif** [comment]
So, lets try to change the test images first, instead of changing the CI tesseract version.

**Comment by lfcnassif** [comment]
tesseract recommends 300dpi images for good OCR results
